node/src/pruner.rs

```diff
@@ -0,0 +1,407 @@
+            blockstore.put_keyed(cid, block.data()).await.unwrap();
+            store
+                .update_sampling_metadata(*height, *status, vec![*cid])
+                .await
+                .unwrap()
+        }
+
+        let pruner = Pruner::start(PrunerArgs {
+            store: store.clone(),
+            blockstore: blockstore.clone(),
+            event_pub: events.publisher(),
+        });
+
+        sleep(Duration::from_secs(1)).await;
+
+        let pruner_event = event_subscriber.recv().await.unwrap().event;
+        assert!(matches!(
+            pruner_event,
+            NodeEvent::PrunedHeaders { to_height: 1_000 }
+        ));
+        assert!(store.get_stored_header_ranges().await.unwrap().is_empty());
+
+        for (height, _, cid, _) in &blocks_with_sampling {
+            assert!(matches!(
+                store.get_sampling_metadata(*height).await.unwrap_err(),
+                StoreError::NotFound
+            ));
+            assert!(!blockstore.has(cid).await.unwrap());
+        }
+
+        pruner.stop();
+
+        assert!(matches!(
+            event_subscriber.try_recv().unwrap_err(),
+            TryRecvError::Empty
+        ));
+    }
+
+    #[async_test]
+    async fn prune_tail() {
+        const BLOCK_TIME: Duration = Duration::from_millis(10);
+
+        let events = EventChannel::new();
+        let store = Arc::new(InMemoryStore::new());
+        let mut gen = ExtendedHeaderGenerator::new();
+        // TODO: insert cids
+        let blockstore = Arc::new(InMemoryBlockstore::new());
+        let mut event_subscriber = events.subscribe();
+
+        // 50 headers before pruning window edge
+        let before_pruning_edge = (Time::now() - (PRUNING_WINDOW + BLOCK_TIME * 100)).unwrap();
+        gen.set_time(before_pruning_edge, BLOCK_TIME);
+        store.insert(gen.next_many_verified(50)).await.unwrap();
+
+        // 10 headers within 1sec of pruning window edge
+        let after_pruning_edge = (Time::now() - (PRUNING_WINDOW - BLOCK_TIME * 10)).unwrap();
+        gen.set_time(after_pruning_edge, BLOCK_TIME);
+        store.insert(gen.next_many_verified(10)).await.unwrap();
+
+        // 50 headers at current time
+        gen.set_time(Time::now(), BLOCK_TIME);
+        store.insert(gen.next_many_verified(10)).await.unwrap();
+
+        let pruner = Pruner::start(PrunerArgs {
+            store: store.clone(),
+            blockstore,
+            event_pub: events.publisher(),
+        });
+
+        sleep(Duration::from_secs(1)).await;
+
+        let pruner_event = event_subscriber.recv().await.unwrap().event;
+        assert!(matches!(
+            pruner_event,
+            NodeEvent::PrunedHeaders { to_height: 50 }
+        ));
+        assert_eq!(
+            store.get_stored_header_ranges().await.unwrap(),
+            new_block_ranges([51..=70])
+        );
+
+        // sleep needs to be longer than PRUNING_INTERVAL for pruner to wake up
+        sleep(Duration::from_secs(12)).await;
```

Comment:
can we add sth like `pruning_interval: Option<Duration>` to `PrunerArgs`, that will only default to the const when `None` is set, to get rid of this sleep?

Reply:
I wasn't sure about parametrising it just for the sake of tests, but ultimately it might be better this way, added :+1: